### PR TITLE
Deprecate the low level SHA* functions.

### PR DESCRIPTION
**Pull request by paulidale**
Use of the low level SHA-1 and SHA-2 functions has been informally discouraged for a long time. We now formally deprecate them.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
Builds with deprecation will fail because `ssl/s3_cbc.c` requires direct access to these functions and it hasn't been moved to a provider.

File: crypto/ct/ct_log.c [hunk 1]
```diff
@@ -7,6 +7,12 @@
  * https://www.openssl.org/source/license.html
  */
 
+/*
+ * SHA256 low level APIs are deprecated for public use, but still ok for
+ * internal use.
+ */
+#include "internal/deprecated.h"
```
Comment (mattcaswell) [inline crypto/ct/ct_log.c]: This file seems to be an example where we *should* convert to using EVP.

File: crypto/ec/curve25519.c [hunk 2]
```diff
@@ -7,6 +7,12 @@
  * https://www.openssl.org/source/license.html
  */
 
+/*
+ * SHA512 low level APIs are deprecated for public use, but still ok for
+ * internal use.
```
Comment (mattcaswell) [inline crypto/ec/curve25519.c]: Seems we should convert this to EVP

File: crypto/ec/ecx_meth.c [hunk 3]
```diff
@@ -7,6 +7,12 @@
  * https://www.openssl.org/source/license.html
  */
 
+/*
+ * SHA-512 low level APIs are deprecated for public use, but still ok for
+ * internal use.
+ */
+#include "internal/deprecated.h"
```
Comment (mattcaswell) [inline crypto/ec/ecx_meth.c]: Convert to EVP?

File: crypto/evp/e_des.c [hunk 4]
```diff
@@ -7,6 +7,12 @@
  * https://www.openssl.org/source/license.html
  */
 
+/*
+ * SHA-1 low level APIs are deprecated for public use, but still ok for
+ * internal use.
+ */
+#include "internal/deprecated.h"
```
Comment (mattcaswell) [inline crypto/evp/e_des.c]: What is the SHA usage in this file that requires this? I couldn't find it.

File: crypto/evp/e_des3.c [hunk 5]
```diff
@@ -7,6 +7,12 @@
  * https://www.openssl.org/source/license.html
  */
 
+/*
+ * SHA-1 low level APIs are deprecated for public use, but still ok for
+ * internal use.
+ */
```
Comment (mattcaswell) [inline crypto/evp/e_des3.c]: Convert to EVP?

File: engines/e_dasync.c [hunk 6]
```diff
@@ -7,6 +7,13 @@
  * https://www.openssl.org/source/license.html
  */
 
+/*
+ * SHA-1 low level APIs are deprecated for public use, but still ok for
+ * internal use.  Note, that due to symbols not being exported, only the
+ * #defines can be accessed, in this case SHA_CBLOCK.
```
Comment (mattcaswell) [inline engines/e_dasync.c]: Perhaps just define SHA_CBLOCK locally in this file rather than including deprecated.h?

File: engines/e_ossltest.c [hunk 7]
```diff
@@ -13,6 +13,13 @@
  * used for any purpose except testing
  */
 
+/*
+ * SHA low level APIs are deprecated for public use, but still ok for
+ * internal use.  Note, that due to symbols not being exported, only the
+ * #defines can be accessed.
+ */
+#include "internal/deprecated.h"
```
Comment (mattcaswell) [inline engines/e_ossltest.c]: Perhaps just define SHA_CBLOCK locally?

**Comment by paulidale** [comment]
CHANGES entry in #10802

File: crypto/evp/e_des.c (hunk 4 above)
Comment (paulidale) [reply crypto/evp/e_des.c]: Me either :(

File: crypto/evp/e_des3.c (hunk 5 above)
Comment (levitte) [reply crypto/evp/e_des3.c]: Wait, what? This is an EVP backend...

Comment (paulidale) [reply crypto/evp/e_des3.c]: Yeah, there is some really crazy stuff in the code if you dig much beyond the superficial.

I was intending to convert this and the others to EVP next week.  Unless a good reason is given for not doing so.  Using EVP will add the cost of a level of function pointer indirection but could gain HW acceleration benefits.

Unless I'm missing something -- which seems to be quite common recently :(

File: engines/e_ossltest.c (hunk 7 above)
Comment (levitte) [reply engines/e_ossltest.c]: I strongly disagree with having that defined "locally" in several places

Comment (levitte) [reply engines/e_ossltest.c]: That is, unless there's justification, such as values can vary depending on location. I assume that's not the case here.

Comment (paulidale) [reply engines/e_ossltest.c]: I'm with @levitte here, best keep the definition in one place.  For the test engine, I don't see a problem with accessing the symbols directly (it needs more than just `SHA_CBLOCK` BTW).  The async engine is example code unless I'm mistaken and it should do the right thing.

I could call `EVP_MD_meth_get_input_blocksize(EVP_sha())` or similar to get the value and set that into the created EVP_MD.  This seems like an all around better approach.

File: crypto/evp/e_des.c (hunk 4 above)
Comment (paulidale) [reply crypto/evp/e_des.c]: I'll take it out and see which CI breaks....

**Comment by vdukhovni** [comment]
> Use of the low level SHA-1 and SHA-2 functions has been informally discouraged for a long time. We now formally deprecate them.

Just deprecate, right?  Not immediately remove...
I rather expect that various external projects that just need a specific hash function, and not much else from OpenSSL or EVP, may be using the underlying functions directly, or at least using EVP_sha1() rather than EVP_get_digestbyname("sha1").

Has anyone checked which projects might be affected? A common source of problems is foreign function bindings in other languages, e.g. crypto, or just hash function modules Perl, Python, Ruby, ... Where the re-exported function is part of the module API.

I suspect that we may not be able to remove these, and perhaps should not, but rather may end up sequestering them in the legacy provider (and/or suitable legacy crypto library where the interface is not EVP and so "providers" don't come into play).  If that's what deprecation means, that's OK. If the intent is to outright remove these functions in a later release, that may prove more difficult.

This is not to say I am opposed to doing this if the downstream impact is not too high, rather this is a note of concern, about perhaps burning too many bridges with the past (chequered as it may be).

**Comment by levitte** [comment]
EVP_sha1() works as a sort of template, causing an implicit fetch of the corresponding provided implementation. This is by design.

Regarding removal of deprecated stuff, it means that the declarations will be removed from public headers, and will no longer be exported by shared libraries. As for the code itself, it will remain internally as long as we need it. If only our providers use it, I suspect it will physically move to that source tree. 

Does that answer your concerns, @vdukhovni ?

**Comment by vdukhovni** [comment]
> Does that answer your concerns, @vdukhovni ?

It clarifies the intent, but the concern is that I don't know the extent to which external projects are then affected by this change. Perhaps old software can continue to use old versions of libcrypto, but the appetite for reworking stable legacy code is generally limited.

If we're providing a general purpose suite of algorithms, even historical ones, it is not clear that the legacy interfaces should be removed. Note again, I'm not saying "no", but raising some concern that perhaps the cost to others may exceed the benefit of doing this, unless there's some specific freedom this gives us to make improvements we could not make otherwise...

File: engines/e_ossltest.c (hunk 7 above)
Comment (paulidale) [reply engines/e_ossltest.c]: All good plans seem to fail.  The engines are also directly accessing sizeof(SHA_CTX) and similar.

Comment (levitte) [reply engines/e_ossltest.c]: So?  the whole ENGINE API will become deprecated as well...  (I gotta finish up that branch!)

Comment (paulidale) [reply engines/e_ossltest.c]: This is better left for the ENGINE deprecation -- I can't see us dropping the either the test or async engine code soon.

I updated the comment to include the `sizeof(ctx)` bit.

File: crypto/evp/e_des3.c (hunk 5 above)
Comment (paulidale) [reply crypto/evp/e_des3.c]: Removed this.

File: engines/e_dasync.c (hunk 6 above)
Comment (paulidale) [reply engines/e_dasync.c]: Added note that `sizeof(ctx)` is also required.

File: crypto/ec/ecx_meth.c (hunk 3 above)
Comment (paulidale) [reply crypto/ec/ecx_meth.c]: Done.

File: crypto/ec/curve25519.c (hunk 2 above)
Comment (paulidale) [reply crypto/ec/curve25519.c]: Done.

**Comment by paulidale** [comment]
Out of WIP.

File: crypto/evp/e_des3.c (hunk 5 above)
Comment (mattcaswell) [reply crypto/evp/e_des3.c]: > Wait, what? This is an EVP backend...

Its an EVP backend for DES not SHA1. It seems reasonable to still use EVP for the SHA-1 bits.

**Comment by mattcaswell** [comment]
> It clarifies the intent, but the concern is that I don't know the extent to which external projects are then affected by this change. Perhaps old software can continue to use old versions of libcrypto, but the appetite for reworking stable legacy code is generally limited.

I'm surprised you bring this up. It has long been our intent to deprecate these low level APIs. At this stage though it is just deprecation. We might get a feel for  how many projects are impacted when 3.0 gets released and we see how many complaints we get about deprecation warnings. We are not tied to any particular timetable for the eventual removal of code - but our stability policy would apply, so this isn't happening any time soon.

File: crypto/ct/ct_log.c (hunk 1 above)
Comment (paulidale) [reply crypto/ct/ct_log.c]: Converted to EVP.

**Comment by paulidale** [comment]
Added MD5 here to reduce later merge conflicts.

**Comment by paulidale** [comment]
Rebased and conflicts resolved (all bar in comments about internal use).

**Comment by paulidale** [comment]
Ping reviewer.

File: test/build.info [hunk 8]
```diff
@@ -601,6 +601,10 @@ IF[{- !$disabled{tests} -}]
     INCLUDE[curve448_internal_test]=.. ../include ../apps/include ../crypto/ec/curve448
     DEPEND[curve448_internal_test]=../libcrypto.a libtestutil.a
 
+    SOURCE[rc4test]=rc4test.c
```
Comment (mattcaswell) [inline test/build.info]: Same comment re internal tests as per the IDEA PR

File: crypto/evp/e_des3.c (hunk 5 above)
Comment (levitte) [reply crypto/evp/e_des3.c]: > Its an EVP backend for DES not SHA1

Goo point.  Okie

**Comment by paulidale** [comment]
> is it the intent to push all commits in this PR, including the "undo" one? Or are we waiting to push this until libssl is fixed so we don't need to have the "undo" commit at all? If the former we need to make sure we have a task somewhere to remind us to come back and fix this properly.

I'm okay either way.  Conflicts will be easier if this is pushed including the "undo" commit but they oughtn't be too terrible.  If this is pushed earlier, I'll create an issue to fix things.

**Comment by mattcaswell** [comment]
> I'm okay either way. Conflicts will be easier if this is pushed including the "undo" commit but they oughtn't be too terrible. If this is pushed earlier, I'll create an issue to fix things.

I suggest we push and create an issue. Otherwise conflicts will be a pain.

File: test/build.info (hunk 8 above)
Comment (paulidale) [reply test/build.info]: Done.

**Comment by paulidale** [comment]
Redo changes are captured in #10877.
If (when) the relevant commit ID changes, I'll update the issue ticket.

**Comment by paulidale** [comment]
Merged to master and undo issue updated.
Thanks for the reviewing and suggestions.

**Pull request closed without merging**